csrc/dynamic_transform.cpp

```diff
@@ -385,6 +648,219 @@ void DynamicTransformConcretizer::concretize() {
   }
 }
 
+void DynamicTransformConcretizer::removeEmptyBranches() {
+  auto fusion = FusionGuard::getCurFusion();
+  for (const auto& empty_tv_descr : info_->getEmptyTensors()) {
+    auto tv = info_->initialInfo()->lookUpTV(empty_tv_descr.tv_name);
+    auto rfactor = TensorDomain::noReductions(tv->getMaybeRFactorDomain());
+    std::vector<Val*> new_shape;
+    new_shape.reserve(rfactor.size());
+    for (auto id : rfactor) {
+      new_shape.push_back(id->getMaybeExpandedExtent());
+    }
+    for (auto ax : empty_tv_descr.empty_axes) {
+      // Hard-code zero extent for empty axes. This lets us detect empty input
+      // and output tensors during scheduling/execution.
+      registerConcretization(new_shape[ax], fusion->zeroVal());
+      new_shape[ax] = fusion->zeroVal();
+    }
+
+    auto hasEmptyRootReductionAxis = [&empty_tv_descr](TensorView* out_tv) {
+      return std::any_of(
+          empty_tv_descr.empty_axes.begin(),
+          empty_tv_descr.empty_axes.end(),
+          [&out_tv](size_t ax) {
+            return out_tv->getRootDomain().at(ax)->isReduction();
```

Comment:
Isn't `ax` an offset within the rfactor no-reduction domain?

Reply:
Right. This is a little confusing. I will change this to refer to rfactor domain instead. In this case, we know there's not an rfactor domain since the op is a reduction, but it will be much clearer to have the descriptor always refer to the maybeRFactor domain.